File: phase-1-arrays/LC49/code.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <sstream>

using namespace std;

class Solution
{
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs)
    {
        unordered_map<string, vector<string>> groups;

        for (const string& s : strs)
        {
            vector<int> count(26, 0);

            for (char c : s)
            {
                count[c - 'a']++;
            }

            ostringstream builder;

            for (int value : count)
            {
                builder << value << '#';
            }

            string key = builder.str();

            groups[key].push_back(s);
        }

        vector<vector<string>> result;

        for (auto& [key, group] : groups)
        {
            result.push_back(group);
        }

        return result;
    }
};

int main()
{
    Solution solution;

    vector<string> strs =
    {
        "eat",
        "tea",
        "tan",
        "ate",
        "nat",
        "bat"
    };

    vector<vector<string>> result =
        solution.groupAnagrams(strs);

    for (const auto& group : result)
    {
        cout << "[";

        for (int i = 0; i < group.size(); i++)
        {
            cout << group[i];

            if (i < group.size() - 1)
                cout << ", ";
        }

        cout << "]" << endl;
    }

    return 0;
}

File: phase-1-arrays/LC49/code.cs
using System;
using System.Collections.Generic;

public class Solution
{
    public IList<IList<string>> GroupAnagrams(string[] strs)
    {
        Dictionary<string, List<string>> groups =
            new Dictionary<string, List<string>>();

        foreach (string s in strs)
        {
            int[] count = new int[26];

            foreach (char c in s)
            {
                count[c - 'a']++;
            }

            string key = string.Join("#", count);

            if (!groups.ContainsKey(key))
            {
                groups[key] = new List<string>();
            }

            groups[key].Add(s);
        }

        IList<IList<string>> result =
            new List<IList<string>>();

        foreach (List<string> group in groups.Values)
        {
            result.Add(group);
        }

        return result;
    }

    public static void Main()
    {
        Solution solution = new Solution();

        string[] strs =
        {
            "eat",
            "tea",
            "tan",
            "ate",
            "nat",
            "bat"
        };

        var result = solution.GroupAnagrams(strs);

        foreach (var group in result)
        {
            Console.WriteLine(
                "[" + string.Join(", ", group) + "]"
            );
        }
    }
}